driver/src/contracts/stablex_contract.rs

```diff
@@ -65,6 +66,10 @@ pub trait StableXContract {
     /// Retrieve the time remaining in the batch.
     fn get_current_auction_remaining_time(&self) -> Result<Duration>;
 
+    /// Searches for the block number of the last block of the given batch. If
+    /// the batch has not yet been finalized, then `None` is returned.
+    fn get_last_block_for_batch(&self, batch_id: u32) -> Result<Option<u64>>;
```

Comment:
I think what we want is `first_block_for_batch`, since in the last block of the previous batch all balances are still wrong (they reflect the balances for the solver in that batch). There might be pending withdraws/deposits that will mature in the next block and should be available to the solver.

This means that there will technically still be a race condition since in the first block of a solving process, there could already be a valid solution applied.

Reply:
Can we use the `getOpenOrderbook` method from the viewer contract then?

Reply:
>  Can we use the getOpenOrderbook method from the viewer contract then?

Yes, I might spend some time today to write a small script which compares the results of `getOpenOrderbookPaginated` on the last block of the previous batch with `getFinalizedOrderbookPaginated` on the first block of the next batch and `getEncodedOrderPaginated` (with client side filtering invalid orders) to get some more confidence in the correctness of these methods (given our issues with them last week).

Reply:
Script is running so we should have data soon. Note, that this also depends on https://github.com/gnosis/dex-contracts/pull/668 in order to reconstruct the orderIds that are required in `submitSolution`